WordPress/gutenberg
# [RNMobile] Add featured image settings to the Latest Posts block

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4034
Partially fixes https://github.com/WordPress/gutenberg/issues/29815

## Related PRs

* `gutenberg-mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/4674

## Description

This PR adds featured image settings to the native Latest Posts block. This is a step towards bringing the available block settings on the native side in alignment with the web. If a user selects the new `Display featured image` toggle, the following two options for customizing the featured image's appearance will be unveiled: 

* Image alignment.
* Add link to featured image.

_Note, the option to specify the featured image's size/dimensions will be addressed separately to this PR. This is because these options are a bit more complex and may warrant some updates to the image block's settings._

_Please refer to the `Types of changes` section below for an overview of the code changes._

## Testing Instructions

### Testing changes in the app

 _Prerequisite: A WordPress site with more than one published blog post, with a featured image assigned to at least one of the posts._

* Open a post or page in the WordPress mobile app and add the Latest Posts block.
* Tap the cog/gear icon to access the block's settings.
* Confirm the new **Featured image settings** section appear within the block's settings and toggle the options.
* Preview the post to confirm that toggling the new settings work as expected i.e. you should be able to view any featured images associated with the latest posts and their alignment should change if adjusted.

### Testing for any web regressions

As these PR introduces some code changes to the web's alignment controls, there is a chance for unintentional regressions. We should therefore test that the alignment controls continue to work as expected while checked out with this branch on the web. Places where the controls appear include within the settings for the Image and Latest Posts blocks.

## Screenshots 

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/157058446-34b9a04d-dcf5-4e67-b6c5-8dfa39fd86fd.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/157943127-4589fa45-9a75-4358-809d-e7e8a5c5ef87.png" width="100%"> |

## Video

https://user-images.githubusercontent.com/2998162/157964542-0a0ca502-3900-4943-bc5b-778b81858356.mov

## Types of changes

New feature (non-breaking change which adds functionality), with the following high-level overview of the code:

* A [new `Featured image settings` section has been added](https://github.com/WordPress/gutenberg/blob/9278d5fdce0842c6032d82596f3311bd800e1100/packages/block-library/src/latest-posts/edit.native.js#L200) to the Latest Posts settings, following the same code pattern used for other sections.
* If a user selects [the new `Display featured image` toggle](https://github.com/WordPress/gutenberg/blob/9278d5fdce0842c6032d82596f3311bd800e1100/packages/block-library/src/latest-posts/edit.native.js#L201), the following two options for customizing the featured image's appearance will be unveiled: 
  * 1. Change the image's alignment. 
  * 2. Add a link to the image. 
* This PR re-uses [the `BlockAlignmentControl` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/alignment-control) to implement the alignment options, which required the following changes:
   * As a significant number of native-specific changes were required, [a native version of the component's ui.js file](https://github.com/WordPress/gutenberg/blob/abf44c8854d19d8726dc6601893ea39e564a884a/packages/block-editor/src/components/block-alignment-control/ui.native.js) was created. The main removal/additions to the original file's code can be found in https://github.com/WordPress/gutenberg/pull/39257/commits/ec59670ae3d7de5aa89c495733de87d418843887.
   * All native-specific parts of the code have been removed from the original ui.js, as they're now redundant. The specific changes can be found in https://github.com/WordPress/gutenberg/pull/39257/commits/40765b409fb2f4e7a401fc18f30e3306a1fa3ef4 and https://github.com/WordPress/gutenberg/pull/39257/commits/9fb7b3893038471152f171170c479df27107a22a.
   * [A new constants.js file](https://github.com/WordPress/gutenberg/blob/abf44c8854d19d8726dc6601893ea39e564a884a/packages/block-editor/src/components/block-alignment-control/constants.js) has been created to add constants that are shared by both the native and web files, such as `BLOCK_ALIGNMENTS_CONTROLS`.  
   * The changes that were made in the native version of `BlockAlignmentControl` enable [the `BottomSheetSelectControl` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/mobile/bottom-sheet-select-control) to be rendered when the `isBottomSheetControl` prop is set to `true`. This was a desired change, as it'll make it easier to add alignment options within a block's bottom sheet settings (rather than only the toolbar) in the future.
   * `BottomSheetSelectControl` was [updated to allow icons](https://github.com/WordPress/gutenberg/blob/21e52a1024810309ca95461c62596baa9f33d946/packages/components/src/mobile/bottom-sheet-select-control/index.native.js#L85) to be displayed for each option in the subsheet, enabling the use of the alignment icons. It's likely that the use of icons will be a future use case for other developers using this component, too.
* _Note, the option to specify the image's size/dimensions will be addressed separately to this PR._ 

## Checklist:
